docs/references/architecture/adr-x-intelligent-validator-peering.md

```diff
@@ -0,0 +1,78 @@
+
+# ADR X: Intelligent Validator Peering
+
+## Changelog
+
+- 11/8: Initial draft
+
+## Status
+
+## Context
+
+The existing CometBFT p2p network is agnostic to the distinction between full nodes and validators. Peers are added via config files and PEX, but validators are equally likely to connect to full nodes as they are to validators via PEX. Thus, validators can be peered suboptimally, requiring messages go through extra hops through full nodes in order for the network to proceed through consensus.
+
+This document proposes a way for validators to peer intelligently with each other, increasing the tendency for validators to be directly peered and thus for consensus messages to travel through the network of validators more quickly. We propose adding a reactor for validator p2p, similar to pex, but specifically for the purpose of getting validators peered with each other.
+
+Note that in this document we will refer to validators as any of the core nodes operated by validator entities in the network. They can be sentry nodes or any non-signing nodes, but should be on the hot-path for consensus messages (i.e. they must participate in the network in order for some consensus messages to be propagated).
+
+## Alternative Approaches
+
+- One alternative for the reactor implementation is complete validator peering, where all validators are peered with all other validators. This is not chosen because it can become prohibitively expensive to a large number of peers, when the number of validators is large.
+
+
+## Decision
+
+## Detailed Design
+
+The CometBFT config should add some new fields which should be populated by validators:
+1. `val_peer_count_low`, `val_peer_count_high`, `val_peer_count_target`. These are the lower bound, upper bound, and target number of validator peers to maintain, respectively.
+2. `isValidator` - whether the node should join the valp2p network.
+
+First, we modify the handshake that nodes perform when connecting to each other. Currently, that handshake exchanges `NodeInfo`. We will augment `NodeInfo` to include an `isValidator` field, which is just the `isValidator` value from the config. This will be kept for each peer of the node in the peer's `NodeInfo`.
```

Comment:
I agree there is potential to spoof into the validator overlay, but I'm not convinced this is any less attack surface than exists now with all node types on one overlay.

But since this is completely new we have the chance to make this a permissioned overlay cross-referenced with the known active set.

https://github.com/cometbft/cometbft/blob/978b84614992cb009b2e37500b6b3a598665a535/state/state.go#L66

Reply:
Oh right, the sentry node case makes this infeasible.

Reply:
i think we should do a simple diffiehellman key exchange so that nodes can verify others in the network. This is the authenticated part of the subnet. Right now its a subnet with a topic that anyone can enter. 

the validator address does not tie to the node address, we would need to change the node address to be based off the consensus key but this would make the assumption that the secret connection needs to work off the consensus key which wouldnt be true if a network uses secp or bls. Changing the secret connection key could be easy enough but when the software moves to quic this will be harder and then we would need some identity authentication system.

Reply:
Do we want this to be actually authenticated? Don't many validators run setups where none of their nodes that have the actual signing key are connected to the network directly?

Reply:
> Changing the secret connection key could be easy enough but when the software moves to quic this will be harder and then we would need some identity authentication system.
Definitely agree that tying things to the existing secret conn implementation is a bad idea. 

> Do we want this to be actually authenticated?
We should figure out what that gives us. If we make this a config option which allows anyone to join the "validator" overlay/subnet, then we can't really make guarantees about the number of validators a node is peered to and so there is a scenario in which these changes don't do anything because everyone sets that config flag.

For sentry nodes, it seems like this is a hacked way of doing what putting validator nodes behind a load balancer does? I'm still undecided on if/how to support that setup if someone uses this new reactor.